mkspiffs: init at 0.2.3

haslersn opened the pull request:
###### Motivation for this change

There wasn't a derivation for mkspiffs yet, so I made one.

Supersedes #46623 as it does, besides the generic version, also contain the versions for arduino-esp32, arduino-esp8266 and esp-idf.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
maintainers: add 'haslersn'

Commit:
mkspiffs: init at 0.2.3

Infinisil commented on pkgs/tools/filesystems/mkspiffs/default.nix:
```diff
@@ -0,0 +1,33 @@
+  version = "0.2.3";
+
+  src = fetchgit {
+    url = "https://github.com/igrr/mkspiffs";
```
Use `fetchFromGitHub` instead

haslersn replied:
fetchFromGitHub doesn't support deepClone as far as I know.

haslersn replied:
However, for `fetchFromGitHub` there is `fetchSubmodules = true;` which does what I want.

Infinisil commented on pkgs/tools/filesystems/mkspiffs/default.nix:
```diff
@@ -0,0 +1,33 @@
+{ stdenv, fetchgit, git
+, extraBuildFlags ? ""
```
Don't take this as an argument

Infinisil commented on pkgs/tools/filesystems/mkspiffs/presets.nix:
```diff
@@ -0,0 +1,24 @@
+let
+  buildMkspiffs = overrides : import ./default.nix (
+    { inherit stdenv fetchgit git; } // overrides
+  );
```
I suggest doing this instead along with the above change:

```nix
{ lib, mkspiffs }:

lib.mapAttrs (name: { extraBuildFlags }:
  (mkspiffs.override {
    buildConfigName = name;
  }).overrideAttrs (drv: {
    buildFlags = drv.buildFlags ++ extraBuildFlags;
  })
) {
  arduino-esp32 = {
    extraBuildFlags = [ "-DSPIFFS_OBJ_META_LEN=4" ];
  };

  # ...
}
```

Infinisil commented on pkgs/tools/filesystems/mkspiffs/default.nix:
```diff
@@ -0,0 +1,33 @@
+  nativeBuildInputs = [ git ];
+  installPhase = ''
+    make clean
+    make dist CPPFLAGS="${extraBuildFlags}" BUILD_CONFIG_NAME="-${buildConfigName}"
```
Isn't the `dist` target the build phase? If so, you should use `buildFlags = [ "dist" ]` instead, along with passing the flags via `makeFlags = [ "CPPFLAGS=..." "..." ];` or via `buildFlags` as well. (buildFlags is applied to only the build phase, while makeFlags to all other make phases as well), while using the install phase for the install only. This should also remove the need to `make clean` (you probably put it there because the default build phase calls `make` without any arguments).

Commit:
fixup! mkspiffs: init at 0.2.3

* Replaced `fetchgit` by `fetchFromGitHub` with `fetchSubmodules = true;`
* Removed the args extraBuildFlags and buildConfigNames from default.nix.
  They used to set the environment variables CPPFLAGS and BUILD_CONFIG_NAME
  which can be achieved by just overriding the corresponding attributes.
* Adapted presets.nix accordingly based on Infinisil's proposal

Infinisil commented:
@GrahamcOfBorg build mkspiffs mkspiffs-presets.arduino-esp8266 mkspiffs-presets.arduino-esp32 mkspiffs-presets.esp-idf

GrahamcOfBorg commented:
<!--REQUEST_ID=58b201dd-a0fd-4d3b-8be8-77f2619fe088-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46674&attempt_id=15587bfe-b355-4607-a251-b8782b034a5c)

Attempted: mkspiffs, mkspiffs-presets.arduino-esp8266, mkspiffs-presets.arduino-esp32, mkspiffs-presets.esp-idf

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking RPATHs of ELF executables and libraries in /nix/store/kvdvw2ky489rjz00ngrzxg0va9kijr5v-mkspiffs-0.2.3
shrinking /nix/store/kvdvw2ky489rjz00ngrzxg0va9kijr5v-mkspiffs-0.2.3/bin/mkspiffs
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/kvdvw2ky489rjz00ngrzxg0va9kijr5v-mkspiffs-0.2.3/bin
patching script interpreter paths in /nix/store/kvdvw2ky489rjz00ngrzxg0va9kijr5v-mkspiffs-0.2.3
checking for references to /build in /nix/store/kvdvw2ky489rjz00ngrzxg0va9kijr5v-mkspiffs-0.2.3...
/nix/store/jxs1rd5n7g1gha3x8d7sfzj8qpi2ga5b-mkspiffs-0.2.3
/nix/store/kvdvw2ky489rjz00ngrzxg0va9kijr5v-mkspiffs-0.2.3
/nix/store/ngx3n33v4lnd6r22phzfds8cpy5qhf7c-mkspiffs-0.2.3
/nix/store/ji2x8i36ywxyxmqwy8wak5m42fykz2b6-mkspiffs-0.2.3
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=58b201dd-a0fd-4d3b-8be8-77f2619fe088-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46674&attempt_id=a4c7c8a5-5fd4-48ea-98cc-db1a63ffebe7)

Attempted: mkspiffs, mkspiffs-presets.arduino-esp8266, mkspiffs-presets.arduino-esp32, mkspiffs-presets.esp-idf

<details><summary>Partial log (click to expand)</summary><p>

```
      _spiffs_object_create in spiffs_nucleus.o
      _spiffs_object_update_index_hdr in spiffs_nucleus.o
  "start", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture i386
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:106: mkspiffs] Error 1
builder for '/nix/store/93qih1v1wmkv812zrsfb2yldfk1398f1-mkspiffs-0.2.3.drv' failed with exit code 2
builder for '/nix/store/x8c30s9216yczz90f0jafab0p20j53dj-mkspiffs-0.2.3.drv' failed with exit code 2
error: build of '/nix/store/145akjfyrcd7nbphr8a0vimg44499iz1-mkspiffs-0.2.3.drv', '/nix/store/93qih1v1wmkv812zrsfb2yldfk1398f1-mkspiffs-0.2.3.drv', '/nix/store/qv3k2rbsg737xcgd6pzl3n9v0bwwwfhy-mkspiffs-0.2.3.drv', '/nix/store/x8c30s9216yczz90f0jafab0p20j53dj-mkspiffs-0.2.3.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=58b201dd-a0fd-4d3b-8be8-77f2619fe088-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46674&attempt_id=3dd4241f-304b-4b95-81ed-fe6e87ce776f)

Attempted: mkspiffs, mkspiffs-presets.arduino-esp8266, mkspiffs-presets.arduino-esp32, mkspiffs-presets.esp-idf

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/d8rcj9miyf6camdwvx87wn74n5xcdlng-mkspiffs-0.2.3/bin
patching script interpreter paths in /nix/store/d8rcj9miyf6camdwvx87wn74n5xcdlng-mkspiffs-0.2.3
patching script interpreter paths in /nix/store/jlvr0smc2kj4f5m6xkc8h7wg1z3rhhyd-mkspiffs-0.2.3
checking for references to /build in /nix/store/jlvr0smc2kj4f5m6xkc8h7wg1z3rhhyd-mkspiffs-0.2.3...
checking for references to /build in /nix/store/d8rcj9miyf6camdwvx87wn74n5xcdlng-mkspiffs-0.2.3...
/nix/store/cr6n6d8qg0kp3lnbjrcyvl7y9ji77vsd-mkspiffs-0.2.3
/nix/store/92y0llfm9pqcv2rp3binal0jdp92l0lf-mkspiffs-0.2.3
/nix/store/d8rcj9miyf6camdwvx87wn74n5xcdlng-mkspiffs-0.2.3
/nix/store/jlvr0smc2kj4f5m6xkc8h7wg1z3rhhyd-mkspiffs-0.2.3
```
</p></details>

Infinisil commented:
If you happen to know how to fix the darwin failure that would be nice. If not, you should disable darwin (maybe set `meta.platforms = platforms.linux`)

Commit:
fixup! mkspiffs: init at 0.2.3

Adapted the comments to the previous fixup

Commit:
fixup! mkspiffs: init at 0.2.3

Added `platforms = platforms.linux` since it doesn't build on darwin

haslersn commented:
> If you happen to know how to fix the darwin failure that would be nice. If not, you should disable darwin (maybe set `meta.platforms = platforms.linux`)

I currently don't know how to fix the failure, so I added the suggested code.

Ekleog commented:
(triage) The concerns from @Infinisil appear to have been taken care of, this looks ready to me. Triggering a build just in case.

@GrahamcOfBorg build mkspiffs mkspiffs-presets.arduino-esp8266 mkspiffs-presets.arduino-esp32 mkspiffs-presets.esp-idf

GrahamcOfBorg commented:
<!--REQUEST_ID=203c77ac-c365-4fa7-a9a6-651b8fa8cfc0-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46674&attempt_id=bcfeaf79-3820-4381-949a-412f07903ce4)

Attempted: mkspiffs, mkspiffs-presets.arduino-esp8266, mkspiffs-presets.arduino-esp32, mkspiffs-presets.esp-idf

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking RPATHs of ELF executables and libraries in /nix/store/g3cbn2bh54xwd06jgl2n0q8a87lsq9nh-mkspiffs-0.2.3
shrinking /nix/store/g3cbn2bh54xwd06jgl2n0q8a87lsq9nh-mkspiffs-0.2.3/bin/mkspiffs
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/g3cbn2bh54xwd06jgl2n0q8a87lsq9nh-mkspiffs-0.2.3/bin
patching script interpreter paths in /nix/store/g3cbn2bh54xwd06jgl2n0q8a87lsq9nh-mkspiffs-0.2.3
checking for references to /build in /nix/store/g3cbn2bh54xwd06jgl2n0q8a87lsq9nh-mkspiffs-0.2.3...
/nix/store/ysp2rshi2qzvhg8kz9ijssh9mi90xsnr-mkspiffs-0.2.3
/nix/store/79z6vn5wk9v1nzrijkin5nqdz8bk13x2-mkspiffs-0.2.3
/nix/store/xiy132mplnjqp3cm4drjlp27zi6wb3iv-mkspiffs-0.2.3
/nix/store/g3cbn2bh54xwd06jgl2n0q8a87lsq9nh-mkspiffs-0.2.3
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=203c77ac-c365-4fa7-a9a6-651b8fa8cfc0-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46674&attempt_id=568291ba-05cd-4e4f-ab28-a9e0f41de19f)

Attempted: mkspiffs, mkspiffs-presets.arduino-esp8266, mkspiffs-presets.arduino-esp32, mkspiffs-presets.esp-idf

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking RPATHs of ELF executables and libraries in /nix/store/qcf1gyingyvfwvw2j80927x1rshfpm5g-mkspiffs-0.2.3
shrinking /nix/store/qcf1gyingyvfwvw2j80927x1rshfpm5g-mkspiffs-0.2.3/bin/mkspiffs
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/qcf1gyingyvfwvw2j80927x1rshfpm5g-mkspiffs-0.2.3/bin
patching script interpreter paths in /nix/store/qcf1gyingyvfwvw2j80927x1rshfpm5g-mkspiffs-0.2.3
checking for references to /build in /nix/store/qcf1gyingyvfwvw2j80927x1rshfpm5g-mkspiffs-0.2.3...
/nix/store/qcf1gyingyvfwvw2j80927x1rshfpm5g-mkspiffs-0.2.3
/nix/store/f6z5g0lcyrfp6nrn30vf48wxgan4njr3-mkspiffs-0.2.3
/nix/store/qrd0k8zv3bfaap0657rskbzvfy97ghkm-mkspiffs-0.2.3
/nix/store/qhikr38fxc141kyam6z02g3p7s8vwikr-mkspiffs-0.2.3
```
</p></details>
